### PR TITLE
CiteSeq Implementation of Standard Workflow

### PR DESCRIPTION
**Pull request by Artemis-R**
addresses issues discussed in #35  and #20 

- updated import functions to support reading of citeseq data
- updated export function to support writing of citeseq data
- updated standard function to perform nearest neighbour calculation to be able to take the number of pcs to use for nearest neighbour calculation as a parameter (allows you to use .X instead of npcs when calculating nn by setting parameter to 0 for citeseq data)
- added function to concatenate adata objects to be able to merge gene expression data with protein values
- added function to normalized citeseq data using clr normalization
- updated standard workflow to support citeseq analysis

### CODE REVIEW

File: besca/_helper.py [hunk 1]
```diff
@@ -190,3 +193,98 @@ def get_means(adata,mycat):
         average_obs=None
         fraction_obs=None
     return(average_obs, fraction_obs)
+
+def concate_adata(adata1, adata2):
```
Comment (hatjek) [inline besca/_helper.py:197]: How is this different from [AnnData.concatenate](https://anndata.readthedocs.io/en/stable/anndata.AnnData.concatenate.html)

Comment (hatjek) [reply besca/_helper.py:197]: @Artemis-R

Comment (Artemis-R) [reply besca/_helper.py:197]: @hatjek From the implementation its similar but it also supports adata objects with .raw objects. This was always an issue for me using the AnnData.concatenate function.

**Comment by Artemis-R** [comment]
Totally agree with @hatjek that the standard workflow will need further alignment. I will add some more things to discuss to the relevant git issue so that we can align on it and then fine-tune the standard pipeline. This version should be a good starting point though.

**Comment by swalpe** [comment]
Checked the cite-seq part, had 3 issues, I think it's a bit of a special case because my data only has 6 proteins. 
**- minor - in the sc.pp.calculate_qc_metrics** I had to change to  e.g. 
qc_adata = sc.pp.calculate_qc_metrics(adata_prot, percent_top=(2, 3, 4), var_type="antibodies", inplace=False)
Perhaps u can add a comment for ppl to be aware they need to adjust here
 **- major - bc.st.pca_neighbors_umap** didn't run for me because of the figure. I tried adding an if statement in the function if(nrpcs_neigh!=0): do the plotting, that seemed to work and the rest ran through; in that function i think we have redundant parameters for the nr. of PCs, can it be? 
**- minor - in concat ** it didn't work because I didn't have the field "CELL" in the adatas. I just added and it work but perhaps we need a check or a warning for that. or do we ever check that on import of the initial object? 

Otherwise all great so I'd suggest to merge once the major point is fixed.

File: besca/_helper.py [hunk 2]
```diff
@@ -190,3 +193,98 @@ def get_means(adata,mycat):
...
+    var_names = adata1.var_names.tolist() + adata2.var_names.tolist()
+
+    if var.index.tolist() != var_names:
+        sys.exit('var and var_names index does not match. Something went wrong.')
+    
+    #create our complete combined object
+    adata_combined = AnnData(X = X, obs = obs, var = var)
+
+    if adata1.raw is not None:
+
```
Comment (ajulienla) [inline besca/_helper.py:245]: Hi,
I might have missed it but what if adata1.raw is None and not adata2.raw ? 
Proposition 
if  any([adata1.raw is not None, adata2.raw is not None]) :
    if any([adata1.raw  is None, adata2.raw is None]):
        sys.exit( ''Only one of the anndata objects contains a .raw!')

Comment (Artemis-R) [reply besca/_helper.py:245]: Agree.

Comment (Artemis-R) [reply besca/_helper.py]: fixed in 613d4dd

**Comment by Artemis-R** [comment]
@swalpe concerning your issues:
- sc.pp.calculate_qc_metrics: this function is dependent on the number of proteins that were measured in the dataset and this of course should not have to be adjusted manually. What do you about changing this function so that it automatically calculates the cumulative values for the top X% of proteins (i.e. if we have 10 proteins then by 10% for 1, if we have 50 then for 5 proteins)
-  bc.st.pca_neighbors_umap: no the parameters for the number of pcs is not redundant. nrpcs indicates how many PCs to calculate. nrpcs_neigh indicates how many of the calculated principle components to use for plotting. Usually they would be equivalent but for citeseq data I wanted to calculate the principle components but then not use them for calculating the nearest neighbours but instead the original values of .X (since we usually do not have ultra-high dimension data). Probably we should incorporate some more checks though for this to adapt it accordingly. I'll open an issue related to this and we can discuss there. 
- concatenate adata: in my opinion we should check for this when we initially import the object. Its been discussed in #12 but has not been implemented yet.

File: besca/_helper.py (hunk 2 above)
Comment (ajulienla) [reply besca/_helper.py]: added just a small commit as there was a typo in the fix i proposed... see ad3477f

**Comment by Artemis-R** [comment]
All points except #12 addressed. This will still need to be fixed separately.

**Pull request merged**